#! /usr/bin/env python
PACKAGE="foot_variables_sync"

from dynamic_reconfigure.parameter_generator_catkin import *
gen = ParameterGenerator()

# Add parameters
#       Name     Type    Level Description     	Default 	Min   	Max

machineState_enum = gen.enum([ gen.const("Homing",int_t, 0, "Homing"),
                       gen.const("Centering", int_t, 1, "Centering"),
                       gen.const("Teleoperation", int_t, 2, "Teleoperation"),
                       gen.const("Emergency", int_t, 3, "Emergency"),
                       gen.const("Standby", int_t, 4, "Standby"),
                       gen.const("Reset", int_t, 5, "Reset"),
                       gen.const("RobotStateControl", int_t, 6, "Robot State Control")],
                     "An enum to set the state machine of the robot")
                     
gen.add ("machine_state", int_t, 0, "Transition state of the machine", 3, 0 , 6, edit_method=machineState_enum)

gen.add ("effortComp_normal", bool_t, 0 , "Enable effort control in the joints",True)
gen.add ("effortComp_constrains", bool_t, 0 , "Enforce soft limits in the workspace",False)
gen.add ("effortComp_compensation", bool_t, 0 , "Enable gravity and friction compensation",False)
gen.add ("effortComp_feedforward", bool_t, 0 , "Enable feedforward components",False)
gen.add ("effortComp_rcmMotion", bool_t, 0 , "Enable rcm control",False)
gen.add ("compensate_leg", bool_t, 0 , "Enable feedforward components",False)

controlledAxis_enum = gen.enum([ gen.const("AllAxis",int_t, -1, "All axis"),
                       gen.const("X", int_t, 0, "First Joint _X"),
                       gen.const("Y", int_t, 1, "Second Joint (Y"),
                       gen.const("PITCH", int_t, 2, "Third Joint (PITCH"),
                       gen.const("ROLL", int_t, 3, "Fourth Joint (ROLL"),
                       gen.const("YAW", int_t, 4, "Fifth Joint (YAW")],
                     "An enum to set the state machine of the robot")
gen.add ("controlled_axis", int_t, 0, "Axis to control", -1, -1 , 4, edit_method=controlledAxis_enum)

controllerType_enum = gen.enum([ gen.const("torque_only",int_t, 0, "TORQUE_ONLY"),
                       gen.const("position_only", int_t, 1, "POSITION_ONLY_or_WS_CONSTRAINS"),
                       gen.const("speed_only", int_t, 2, "SPEED_ONLY_or_MOTION_DAMPING"),
                       gen.const("position_speed_cascade", int_t, 3, "POSITION_SPEED_CASCADE_or_BOTH_CONSTRAINS"),
                       gen.const("speed_position_cascade", int_t, 4, "SPEED_POSITION_CASCADE")],
                     "An enum to set controller type for the robot")
gen.add ("controller_type", int_t, 0, "Controllery type of the robot", 0, 0 , 6, edit_method=controllerType_enum)


gen.add ("send_this_position",bool_t,0, "Use these positions for position control", False)
gen.add ("send_zero_effort",bool_t,0, "Send zero desired torque to the platform", False)
gen.add ("capture_platform_position",bool_t,0, "Update the dynamic reconfiguration with current platform position", False)
gen.add("desired_Position_X", double_t,0,"X axis position [m]",0.0,-0.095,0.095)
gen.add("desired_Position_Y", double_t,0,"Y axis position [m]",0.0,-0.087,0.087)
gen.add("desired_Position_PITCH", double_t,0,"PITCH axis position [deg]",0.0,-20.0,20.0)
gen.add("desired_Position_ROLL", double_t,0,"ROLL axis position [deg]",0.0,-20.0,20.0)
gen.add("desired_Position_YAW", double_t,0,"YAW axis position deg]",0.0,-25.0,25.0)

gen.add ("use_default_gains",bool_t,0, "Use the default gains from the platform", True)
gen.add("kp_Position_X", double_t,0,"KP position X [N/m]",0.0,0.0,15000)
gen.add("kp_Position_Y", double_t,0,"KP position Y  [N/m]",0.0,0.0,15000)
gen.add("kp_Position_PITCH", double_t,0,"KP position PITCH _x0.0001_[ Nm/rad]",0.0,0.0,15000)
gen.add("kp_Position_ROLL", double_t,0,"KP position POLL _x0.0001_[ Nm/rad]",0.0,0.0,15000)
gen.add("kp_Position_YAW", double_t,0,"KP position YAW _x0.0001_[ Nm/rad]",0.0,0.0,15000)

gen.add("ki_Position_X", double_t,0,"KI position X [N/m.s]",0.0,0.0,15000.0)
gen.add("ki_Position_Y", double_t,0,"KI position Y [N.s/m.s]",0.0,0.0,15000.0)
gen.add("ki_Position_PITCH", double_t,0,"KI position PITCH _x0.0001_[ Nm/rad.s]",0.0,0.0,15000.0)
gen.add("ki_Position_ROLL", double_t,0,"KI position ROLL _x0.0001_[ Nm/rad.s]",0.0,0.0,15000.0)
gen.add("ki_Position_YAW", double_t,0,"KI position YAW _x0.0001_[ Nm/rad.s]",0.0,0.0,15000.0)

gen.add("kd_Position_X", double_t,0,"KD position X [N.s/m]",0.0,0.0,60.0)
gen.add("kd_Position_Y", double_t,0,"KD position Y [N.s/m]",0.0,0.0,60.0)
gen.add("kd_Position_PITCH", double_t,0,"KD position PITCH _x0.0001_[Nm.s/rad]",0.0,0.0,60.0)
gen.add("kd_Position_ROLL", double_t,0,"KD position ROLL _x0.0001_[ Nm.s/rad]",0.0,0.0,60.0)
gen.add("kd_Position_YAW", double_t,0,"KD position YAW _x0.0001_[ Nm.s/rad]",0.0,0.0,60.0)

gen.add("kp_Speed_X", double_t,0,"KP speed X _x0.01_[N.s/m]",0.0,0.0,2000.0)
gen.add("kp_Speed_Y", double_t,0,"KP speed Y _x0.01_[N.s/m]",0.0,0.0,2000.0)
gen.add("kp_Speed_PITCH", double_t,0,"KP speed PITCH _x0.00001_[Nm.s/rad]",0.0,0.0,1500)
gen.add("kp_Speed_ROLL", double_t,0,"KP speed ROLL _x0.00001_[Nm.s/rad]",0.0,0.0,1500)
gen.add("kp_Speed_YAW", double_t,0,"KP speed YAW _x0.00001_[Nm.s/rad]",0.0,0.0,1500)

gen.add("ki_Speed_X", double_t,0,"KI speed X _x0.01_[N/m]",0.0,0.0,3000)
gen.add("ki_Speed_Y", double_t,0,"KI speed Y _x0.01_[N.s/m2.s]",0.0,0.0,3000)
gen.add("ki_Speed_PITCH", double_t,0,"KI speed PITCH _x0.00001_[Nm/rad2.s]",0.0,0.0,7000.0)
gen.add("ki_Speed_ROLL", double_t,0,"KI speed ROLL _x0.00001_[Nm/rad2.s]",0.0,0.0,7000.0)
gen.add("ki_Speed_YAW", double_t,0,"KI speed YAW _x0.00001_[Nm/rad2.s]",0.0,0.0,7000.0)

gen.add("kd_Speed_X", double_t,0,"KD speed X _x0.01_[N.s/m]",0.0,0.0,2.0)
gen.add("kd_Speed_Y", double_t,0,"KD speed Y _x0.01_[N.s/m]",0.0,0.0,1.0)
gen.add("kd_Speed_PITCH", double_t,0,"KD speed PITCH _x0.00001_[Nm.s/rad]",0.0,0.0,10.0)
gen.add("kd_Speed_ROLL", double_t,0,"KD speed  ROLL _x0.00001_[Nm.s/rad]",0.0,0.0,10.0)
gen.add("kd_Speed_YAW", double_t,0,"KD  speed YAW _x0.00001_[Nm.s/rad]",0.0,0.0,10.0)

exit(gen.generate(PACKAGE, "foot_variables_sync", "machineStateParams"))
